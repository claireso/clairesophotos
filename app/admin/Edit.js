import React from 'react'
import Form from './form/Form'

export default ({ photo }) => {
  return (
    <main>
      <h1>Edit photo</h1>
      <Form photo={ photo } />
    </main>
  )
}
