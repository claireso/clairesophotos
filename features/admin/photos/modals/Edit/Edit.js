import { Fragment, useEffect, useCallback, memo } from 'react'
import PropTypes from 'prop-types'

import { Heading1 } from '@components/Headings'

import Form from '../components/Form'

import PhotosReducer from '../../reducer'

const Edit = ({ id }) => {
  const [state, { loadResource, editResource }] =
    PhotosReducer.usePhotosReducer()

  const isProcessing = state.status === 'pending'

  let photo = state.items.find((p) => p.id === id)

  if (!photo) {
    photo = state.single?.id === id ? state.single : undefined
  }

  useEffect(() => {
    if (!photo) {
      loadResource(id)
    }
  }, [id, photo, loadResource])

  const onSubmit = useCallback(
    (data) => {
      editResource(id, data)
    },
    [id, editResource]
  )

  if (photo === undefined) return null

  return (
    <Fragment>
      <Heading1>Edit photo</Heading1>
      <Form onSubmit={onSubmit} photo={photo} isProcessing={isProcessing} />
    </Fragment>
  )
}

Edit.propTypes = {
  id: PropTypes.number.isRequired
}

export default memo(Edit)
