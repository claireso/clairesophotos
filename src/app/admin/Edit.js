import React from 'react'
import PropTypes from 'prop-types'

import Page from './components/Page'
import Form from './form/Form'
import Toolbar from './components/Toolbar'
import { BackLink } from './components/Links'

const Edit = ({ photo }) => {
  return (
    <Page title="Edit photo">
      <Toolbar>
        <BackLink href="/admin/photos" label="Back to list" />
      </Toolbar>
      <Form photo={photo} />
    </Page>
  )
}

Edit.propTypes = {
  photo: PropTypes.object.isRequired
}

export default Edit
