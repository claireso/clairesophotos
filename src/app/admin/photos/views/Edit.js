import React from 'react'
import PropTypes from 'prop-types'

import Modal from '../../components/Modal'
import Flash from '../../components/Flash'

import Form from './form/Form'

class Edit extends React.Component {
  componentDidMount() {
    if (!this.props.photo) {
      this.props.loadPhoto(this.props.id)
    }
  }

  render() {
    const { photo, error } = this.props

    if (photo === undefined) return null

    return (
      <Modal onClose={() => this.props.navigate('/admin/photos')}>
        {error && <Flash {...error} />}
        <h1>Edit photo</h1>
        <Form
          onSubmit={this.props.editPhoto.bind(this, photo.id)}
          photo={photo}
        />
      </Modal>
    )
  }
}

Edit.propTypes = {
  id: PropTypes.string.isRequired,
  navigate: PropTypes.func.isRequired,
  error: PropTypes.shape({
    status: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired
  }),
  editPhoto: PropTypes.func.isRequired,
  loadPhoto: PropTypes.func.isRequired,
  photo: PropTypes.object
}

export default Edit
