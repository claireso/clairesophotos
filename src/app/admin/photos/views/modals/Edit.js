import React, { Fragment } from 'react'
import PropTypes from 'prop-types'

import Flash from '@admin/components/Flash'
import { Heading1 } from '@admin/components/Headings'

import Form from './form/Form'

class Edit extends React.PureComponent {
  componentDidMount() {
    if (!this.props.photo) {
      this.props.loadPhoto(this.props.id)
    }
  }

  render() {
    const { photo, error, isProcessing } = this.props

    if (photo === undefined) return null

    return (
      <Fragment>
        {error && <Flash {...error} />}
        <Heading1>Edit photo</Heading1>
        <Form
          onSubmit={this.props.editPhoto.bind(this, photo.id)}
          photo={photo}
          isProcessing={isProcessing}
        />
      </Fragment>
    )
  }
}

Edit.propTypes = {
  error: PropTypes.shape({
    status: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired
  }),
  editPhoto: PropTypes.func.isRequired,
  loadPhoto: PropTypes.func.isRequired,
  photo: PropTypes.object,
  isProcessing: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  id: PropTypes.number.isRequired
}

export default Edit
