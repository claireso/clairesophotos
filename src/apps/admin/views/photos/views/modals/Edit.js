import React, { Fragment, useEffect, useCallback } from 'react'
import PropTypes from 'prop-types'

import Flash from '@admin/components/Flash'
import { Heading1 } from '@admin/components/Headings'

import Form from './form/Form'

const Edit = ({ id, photo, error, isProcessing, loadPhoto, editPhoto }) => {
  useEffect(() => {
    if (!photo) {
      loadPhoto(id)
    }
  }, [id, loadPhoto, photo])

  const onSubmit = useCallback(
    data => {
      editPhoto(id, data)
    },
    [editPhoto, id]
  )

  if (photo === undefined) return null

  return (
    <Fragment>
      {error && <Flash {...error} />}
      <Heading1>Edit photo</Heading1>
      <Form onSubmit={onSubmit} photo={photo} isProcessing={isProcessing} />
    </Fragment>
  )
}

Edit.propTypes = {
  error: PropTypes.shape({
    status: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired
  }),
  editPhoto: PropTypes.func.isRequired,
  loadPhoto: PropTypes.func.isRequired,
  photo: PropTypes.object,
  isProcessing: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  id: PropTypes.number.isRequired
}

export default Edit
