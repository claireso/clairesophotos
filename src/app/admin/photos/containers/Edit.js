import { connect } from 'react-redux'
import { navigate } from '@reach/router'

import Edit from '../views/Edit'

import {
  editPhoto,
  EDIT_PHOTO_SUCCESS,
  loadPhoto
} from '@common/actions/photos'

const mapStateToProps = (state, props) => ({
  photo:
    state.photos.items.find(photo => photo.id === Number(props.id)) ||
    state.photos.detail,
  error: state.photos.error
})

const mapDispatchToProps = dispatch => ({
  editPhoto(id, data) {
    dispatch(editPhoto(id, data)).then(action => {
      if (action.type === EDIT_PHOTO_SUCCESS) {
        navigate('/admin/photos')
        return
      }

      document.querySelector('.modal').scrollTo(0, 0)
    })
  },

  loadPhoto(id) {
    dispatch(loadPhoto(id))
  }
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Edit)
