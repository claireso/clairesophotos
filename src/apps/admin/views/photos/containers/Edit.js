import { connect } from 'react-redux'
import { navigate } from '@reach/router'
import qs from 'qs'

import Edit from '../views/modals/Edit'

import { editPhoto, EDIT_PHOTO_SUCCESS, loadPhoto } from '@admin/actions/photos'

import { displaySuccessMessage } from '@admin/actions/messages'

const mapStateToProps = state => ({
  error: state.photos.error,
  isProcessing: state.photos.isProcessing
})

const mapDispatchToProps = dispatch => ({
  editPhoto(id, data) {
    dispatch(editPhoto(id, data)).then(action => {
      if (action.type === EDIT_PHOTO_SUCCESS) {
        const query = qs.parse(window.location.search.substring(1))
        const search = qs.stringify({
          ...query,
          action: undefined,
          id: undefined
        })
        navigate(`?${search}`)
        dispatch(
          displaySuccessMessage({
            message: 'Your photo has been updated successfully',
            key: 'CRUD_PHOTO'
          })
        )
        window.scrollTo(0, 0)
        return
      }

      document.querySelector('#modal').scrollTo(0, 0)
    })
  },

  loadPhoto(id) {
    dispatch(loadPhoto(id))
  }
})

export default connect(mapStateToProps, mapDispatchToProps)(Edit)
