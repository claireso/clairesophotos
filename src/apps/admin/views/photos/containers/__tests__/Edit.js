import React from 'react'
import { Provider } from 'react-redux'
import { render, fireEvent, wait } from '@testing-library/react'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import Edit from '../Edit'
import callApiMiddleware from '@admin/middleware/callApi'

jest.mock('axios', () => {
  return jest.fn().mockImplementation(({ url }) => {
    switch (url) {
      case '/photos/1':
        return Promise.resolve({
          data: {
            id: 1,
            title: 'Single photography edit',
            description: 'Janvier 2019',
            name: '01d2tf2h38pwcd953ans2f64p7.jpg',
            position: 'center',
            portrait: true,
            square: false,
            created_at: '2019-02-03T19:59:00.088Z',
            updated_at: '2019-02-03T19:59:00.088Z'
          }
        })

      default:
        return Promise.resolve({ data: {} })
    }
  })
})

const middlewares = [thunk, callApiMiddleware]
const mockStore = configureMockStore(middlewares)

describe('<Edit />', () => {
  const renderComponent = (props = {}) => {
    const store = mockStore({ user: {}, photos: { isProcessing: false } })

    const rendered = render(
      <Provider store={store}>
        <Edit onClose={() => {}} {...props} id={1} />
      </Provider>
    )

    return {
      ...rendered,
      store
    }
  }

  test('should render component', () => {
    const { container } = renderComponent({ photo: global.__PHOTO__ })

    expect(container).toMatchSnapshot()
  })

  test('should load photo', async () => {
    const { container, store } = renderComponent()

    expect(container).toMatchSnapshot()

    const expectedActions = [
      { type: 'LOAD_PHOTO_REQUEST' },
      {
        type: 'LOAD_PHOTO_SUCCESS',
        response: {
          id: 1,
          title: 'Single photography edit',
          description: 'Janvier 2019',
          name: '01d2tf2h38pwcd953ans2f64p7.jpg',
          position: 'center',
          portrait: true,
          square: false,
          created_at: '2019-02-03T19:59:00.088Z',
          updated_at: '2019-02-03T19:59:00.088Z'
        }
      }
    ]

    await wait(() => {
      expect(store.getActions()).toEqual(expectedActions)
    })
  })

  test('should edit photo', async () => {
    const { container, getByLabelText, store } = renderComponent({
      photo: global.__PHOTO__
    })

    fireEvent.change(getByLabelText(/title/i), {
      target: {
        value: 'Single photography edit'
      }
    })

    fireEvent.submit(container.querySelector('form'))

    const expectedActions = [
      { type: 'EDIT_PHOTO_REQUEST' },
      {
        type: 'EDIT_PHOTO_SUCCESS',
        response: {
          id: 1,
          title: 'Single photography edit',
          description: 'Janvier 2019',
          name: '01d2tf2h38pwcd953ans2f64p7.jpg',
          position: 'center',
          portrait: true,
          square: false,
          created_at: '2019-02-03T19:59:00.088Z',
          updated_at: '2019-02-03T19:59:00.088Z'
        }
      },
      {
        type: 'ADD_MESSAGE',
        status: 'success',
        key: 'CRUD_PHOTO',
        message: 'Your photo has been updated successfully'
      }
    ]

    await wait(() => {
      expect(store.getActions()).toEqual(expectedActions)
    })
  })
})
