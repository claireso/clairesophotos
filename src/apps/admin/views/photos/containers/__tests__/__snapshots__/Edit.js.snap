// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Edit /> should load photo 1`] = `<div />`;

exports[`<Edit /> should render component 1`] = `
.c0 {
  font-size: 2.4rem;
  margin: 0 0 3rem;
}

.c1 {
  margin: 0 0 var(--gutter);
}

.c8 {
  margin: 0 0 var(--gutter);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c8 > label {
  margin: 0 1rem 0 0;
}

.c2 {
  display: block;
  font-size: 1.6rem;
  margin: 0 0 1rem;
}

.c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 1px solid var(--gray-1);
  border-radius: 0.4rem;
  box-shadow: 1px 1px 3px var(--box-shadow);
  display: block;
  outline: none;
  padding: 1rem;
  width: 100%;
  font-size: 1.4rem;
  -webkit-transition: border-color 250ms ease-out;
  transition: border-color 250ms ease-out;
}

.c3:focus {
  border-color: var(--gray-2);
}

.c9 {
  position: relative;
}

.c9 svg {
  position: absolute;
  top: 50%;
  right: 0.9rem;
  bottom: 0;
  -webkit-transform: translate(0,-50%) rotate(90deg);
  -ms-transform: translate(0,-50%) rotate(90deg);
  transform: translate(0,-50%) rotate(90deg);
  pointer-events: none;
}

.c10 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: var(--white);
  border: 1px solid var(--gray-1);
  border-radius: 0.5rem;
  box-shadow: 1px 1px 3px var(--box-shadow);
  cursor: pointer;
  font-family: 'Roboto',Arial,sans-serif;
  font-size: 1.4rem;
  outline: none;
  padding: 0.8rem 2.6rem 0.8rem 0.8rem;
}

.c10:focus {
  border-color: var(--gray-2);
}

.c4 {
  border: 1px solid var(--gray-1);
  border-radius: 0.4rem;
  box-shadow: 1px 1px 3px var(--box-shadow);
  padding: 2rem;
  position: relative;
}

.c5 img {
  display: block;
  max-width: 100%;
  max-height: 17rem;
  margin: 0 auto var(--gutter);
}

.c6 svg {
  display: block;
  width: 2.4rem;
  height: 2.4rem;
  fill: var(--primary);
  margin: 0 auto;
}

.c6 span {
  display: block;
  font-size: 1.4rem;
  margin: 1rem 0 0;
  text-align: center;
}

.c6 small {
  color: var(--gray-2);
  display: block;
  font-size: 1.2rem;
  margin: 0.5rem 0 0;
  text-align: center;
}

.c7 {
  bottom: 0;
  cursor: pointer;
  left: 0;
  opacity: 0;
  position: absolute;
  right: 0;
  top: 0;
  width: 100%;
}

.c11 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: var(--primary);
  border: none;
  border-radius: 3.6rem;
  color: white;
  cursor: pointer;
  display: block;
  font-size: 1.4rem;
  margin: 3rem auto 0;
  outline: none;
  padding: 1.5rem 5.5rem;
  -webkit-transition: background 150ms ease-out;
  transition: background 150ms ease-out;
}

.c11:hover {
  background: var(--primary-lighter);
}

<div>
  <h1
    class="c0"
  >
    Edit photo
  </h1>
  <form
    action=""
    enctype="multipart/form-data"
    method="POST"
  >
    <div
      class="c1"
    >
      <label
        class="c2"
        for="title"
      >
        Title
      </label>
      <input
        class="c3"
        id="title"
        name="title"
        type="text"
        value="Single photography"
      />
    </div>
    <div
      class="c1"
    >
      <label
        class="c2"
        for="description"
      >
        Description
      </label>
      <input
        class="c3"
        id="description"
        name="description"
        type="text"
        value="Janvier 2019"
      />
    </div>
    <div
      class="c1"
    >
      <label
        class="c2"
        for="file"
      >
        Photo
      </label>
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <img
            src="/img/01d2tf2h38pwcd953ans2f64p7.jpg"
          />
        </div>
        <div
          class="c6"
        >
          <svg
            height="16"
            viewBox="0 0 26 28"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 23c0-0.547-0.453-1-1-1s-1 0.453-1 1 0.453 1 1 1 1-0.453 1-1zM24 23c0-0.547-0.453-1-1-1s-1 0.453-1 1 0.453 1 1 1 1-0.453 1-1zM26 19.5v5c0 0.828-0.672 1.5-1.5 1.5h-23c-0.828 0-1.5-0.672-1.5-1.5v-5c0-0.828 0.672-1.5 1.5-1.5h6.672c0.422 1.156 1.531 2 2.828 2h4c1.297 0 2.406-0.844 2.828-2h6.672c0.828 0 1.5 0.672 1.5 1.5zM20.922 9.375c-0.156 0.375-0.516 0.625-0.922 0.625h-4v7c0 0.547-0.453 1-1 1h-4c-0.547 0-1-0.453-1-1v-7h-4c-0.406 0-0.766-0.25-0.922-0.625-0.156-0.359-0.078-0.797 0.219-1.078l7-7c0.187-0.203 0.453-0.297 0.703-0.297s0.516 0.094 0.703 0.297l7 7c0.297 0.281 0.375 0.719 0.219 1.078z"
            />
          </svg>
          <span>
            Upload new photo
          </span>
          <small>
            (only jpg and png)
          </small>
          <input
            class="c7"
            name="file"
            type="file"
          />
        </div>
      </div>
    </div>
    <div
      class="c8"
    >
      <label
        class="c2"
        for="position"
      >
        Position
      </label>
      <div
        class="c9"
      >
        <select
          class="c10"
          id="position"
          name="position"
          required=""
        >
          <option
            value="left"
          >
            Left
          </option>
          <option
            selected=""
            value="center"
          >
            Center
          </option>
          <option
            value="right"
          >
            Right
          </option>
        </select>
        <svg
          height="14"
          viewBox="0 0 9 28"
          width="9"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.297 15c0 0.125-0.063 0.266-0.156 0.359l-7.281 7.281c-0.094 0.094-0.234 0.156-0.359 0.156s-0.266-0.063-0.359-0.156l-0.781-0.781c-0.094-0.094-0.156-0.219-0.156-0.359 0-0.125 0.063-0.266 0.156-0.359l6.141-6.141-6.141-6.141c-0.094-0.094-0.156-0.234-0.156-0.359s0.063-0.266 0.156-0.359l0.781-0.781c0.094-0.094 0.234-0.156 0.359-0.156s0.266 0.063 0.359 0.156l7.281 7.281c0.094 0.094 0.156 0.234 0.156 0.359z"
          />
        </svg>
      </div>
    </div>
    <input
      class="c11"
      type="submit"
      value="Save"
    />
  </form>
</div>
`;
